---
title: "Databases used for testing and training ampir"
output: github_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, message = FALSE, warning = FALSE)
```

```{r}
library(ampir)
library(tidyverse)
library(ggpubr)
```

## Public AMP databases

The following databases were accessed as of March 2020. All `ampir` models were trained using data from one or more of these databases (see details below). 

Four recently updated antimicrobial peptide (AMP) databases were used:

* [APD 3](http://aps.unmc.edu/AP/) by [Wang et al. 2016](https://academic.oup.com/nar/article/44/D1/D1087/2503090) 
* [DRAMP 2.0](http://dramp.cpu-bioinfor.org/) by [Kang et al. 2019](https://www.ncbi.nlm.nih.gov/pubmed/31409791).
* [dbAMP](http://140.138.77.240/~dbamp/index.php) by [Jhong et al. 2018](https://www.ncbi.nlm.nih.gov/pubmed/30380085)
* [UniProt](https://www.uniprot.org/uniprot/?query=keyword%3A%22Antimicrobial+%5BKW-0929%5D%22&sort=score) using the keyword "Antimicrobial [KW-0929]".

Raw downloads for these databases are included in the data distribution. After unpacking they should be present at the following file locations:

| Database Name | File |
| ------------------------ | --------------- |
| APD 3 | `raw_data/amp_databases/APD_032020.xlsx` |
| DRAMP Natural | `raw_data/amp_databases/dramp_nat_tidy.fasta` |
| dbAMP | `raw_data/amp_databases/dbAMPv1.4.xlsx` |
| UniProt | `raw_data/amp_databases/uniprot-keyword__Antimicrobial+[KW-0929]_.xlsx` |



### APD

```{r}
apd <- readxl::read_excel("raw_data/amp_databases/APD_032020.xlsx", col_names = c("ID","description","seq")) %>% 
  mutate(uniprot = str_match(description,"(^[A-Z][0-9][A-Z,0-9]*),")[,2]) %>% 
  mutate(length = nchar(seq)) %>% 
  add_column(database = "APD")
```


The [Antimicrobial Peptide Database](http://aps.unmc.edu/AP/main.php) appears to be regularly updated and contained `r nrow(apd)` AMPs in March 2020. An outdated (2017) AMP sequence list is downloadable from http://aps.unmc.edu/AP/downloads.php which currently contains 2,338 sequences. To include the updated AMP list, the [web query interface](http://aps.unmc.edu/AP/database/query_input.php) was used to obtain the full 3,177 AMPs along with sufficient metadata to filter unwanted entries. 

### DRAMP

```{r}
dramp <- read_faa("raw_data/amp_databases/dramp_nat_tidy.fasta") %>% 
  separate(seq_name,into = c("ID","description"),sep = "\\|") %>% 
  add_column(uniprot = NA) %>% 
  mutate(length = nchar(seq_aa)) %>% 
  rename(seq = seq_aa) %>% 
  add_column(database = "DRAMP")
```


[DRAMP's download page](http://dramp.cpu-bioinfor.org/downloads/) provides access to a general AMP dataset (which contains both natural and synthetic AMPs). This general dataset was posted on 06/08/2019 but according to the "news and events" section, the [natural dataset](http://dramp.cpu-bioinfor.org/browse/NaturalData.php) has been updated several times since then. Because the natural dataset contains the AMPs we are interested in and is also more regularly updated (it currently contains `r nrow(dramp)` sequences), the natural AMP data was obtained using a [scrape script](scripts/scrape_dramp.sh).


### dbAMP 

```{r}
dbamp <- readxl::read_excel("raw_data/amp_databases/dbAMPv1.4.xlsx") %>% 
  filter(!grepl('Synthetic|synthetic', dbAMP_Taxonomy)) %>%
  select(ID = dbAMP_ID, description = Activity, seq = Sequence, uniprot = dbAMP_UniProt,length=Length) %>%
  add_column(database = "dbAMP")

# Note that the reported length in dbamp does not always match the number of characters in the amino acid sequence.
# We fix the length field to reflect this
dbamp <- dbamp %>% mutate(length = nchar(seq))
```


The latest release of dbAMP is from 06/2019 and was downloaded from their [download page](http://140.138.77.240/~dbamp/download.php). It contains `r nrow(dbamp)` experimentally verified natural AMPs (synthetic AMPs were removed).

```{bash, include=FALSE, eval = FALSE}
wget http://140.138.77.240/~dbamp/download/dbAMPv1.4.tar.gz -O raw_data/amp_databases/dbAMPv1.4.tar.gz
tar -xf raw_data/amp_databases/dbAMPv1.4.tar.gz -C raw_data/amp_databases
```

### UniProt

```{r}
uniprot <- readxl::read_excel("raw_data/amp_databases/uniprot-keyword__Antimicrobial+[KW-0929]_.xlsx")

swissprot <- uniprot %>% 
  filter(Status == "reviewed") %>% 
  rename(uniprot = Entry) %>% 
  select(ID = `Entry name`,description = `Protein names`, seq = Sequence,uniprot, length = Length) %>% 
  add_column(database = "SwissProt")
```


AMPs were downloaded from UniProt on 14-April-2020 using the search term: "keyword:Antimicrobial [KW-0929]".  This included `r uniprot %>% filter(Status=="reviewed") %>% nrow()` reviewed and `r uniprot %>% filter(Status=="unreviewed") %>% nrow()` unreviewed proteins.

```{r}
combined <- rbind(apd,dramp,dbamp,swissprot)
```

## Summary of AMP databases

One of the most striking differences between AMP databases becomes clear simply by looking at the length distributions. The `APD` and `DRAMP` databases emphasise short peptides (mostly < 50 amino acids) which reflects their focus on mature peptides rather than full length precursor proteins.

```{r}
ggplot(combined,aes(x=length)) + geom_histogram(binwidth = 3) +
  facet_wrap(~database, scales = "free_y", ncol = 1) +
  labs(x = "Sequence length", y = "Frequency") + xlim(0,300) +
  theme_pubclean()
ggsave("figures/amp_lengths.png", height = 6, width = 6)
```



```{r}
uniprot_peptides <- uniprot %>% 
  mutate(pep_start = as.integer(str_match(Peptide, "PEPTIDE.*?([0-9]+)")[,2])) %>% 
  mutate(pep_end = as.integer(str_match(Peptide, "PEPTIDE.*?([0-9]+)..([0-9]+)")[,3])) %>% 
  mutate(pep_length = pep_end - pep_start + 1) %>% 
  mutate(is_mature_peptide = abs(pep_length - Length)<3) %>% 
  mutate(signal = !is.na(`Signal peptide`)) %>% 
  mutate(maturity = ifelse(is.na(is_mature_peptide),"Unknown",is_mature_peptide)) %>% 
  mutate(maturity = ifelse(maturity==FALSE,"Precursor",maturity)) %>% 
  mutate(maturity = ifelse(maturity==TRUE,"Mature",maturity))
```

The SwissProt database provides a `Peptide` field which allows us to distinguish between entries for mature peptides and precursors. If the Peptide length is the same as the total length, it is a mature peptide. For entries where such information is available we see a very clear split with mature peptides having a typical length of 25 amino acids (AA) whereas precursors are slightly longer at around 60-75 AA. Note that the length distribution of mature peptides matches the APD length distribution very well but all other databases, including dbAMP include longer sequences and therefore are likely to include some fraction of precursors. Note that there are a total of `r uniprot_peptides %>% filter(is_mature_peptide) %>% nrow()` mature peptides, `r uniprot_peptides %>% filter(!is_mature_peptide) %>% nrow()` precursors with peptide annotation information, and `r uniprot_peptides %>% filter(is.na(is_mature_peptide)) %>% filter(Status=="reviewed") %>% nrow()` reviewed AMPs without Peptide information. The `Unknown` category has a notably broader distribution of lengths reflecting the possibility that it includes a mix of both types.  

```{r, warning=FALSE, message=FALSE}
uniprot_peptides %>% 
  filter(Status=="reviewed") %>%
  mutate(type = ifelse(is.na(is_mature_peptide),"Unknown",is_mature_peptide)) %>% 
  mutate(type = ifelse(type=="TRUE","Mature",type)) %>% 
  mutate(type = ifelse(type=="FALSE","Precursor",type)) %>% 
  ggplot(aes(x=Length)) + geom_density(aes(color=type)) + 
  xlim(0,300) + theme_pubclean() +
  theme(legend.title = element_blank()) +
  labs(x= "Sequence length", y = "Density")
ggsave("figures/swissprot_lengths.png", height = 4, width = 8)
```


```{r}
swissprot_mature <- uniprot_peptides %>% filter(Status=="reviewed") %>% filter(is_mature_peptide) %>% mutate(signal = !is.na(`Signal peptide`))
swissprot_precursor <- uniprot_peptides %>% filter(Status=="reviewed") %>% filter(!is_mature_peptide) %>% mutate(signal = !is.na(`Signal peptide`))
swissprot_unknown <- uniprot_peptides %>% filter(Status=="reviewed") %>% filter(is.na(is_mature_peptide)) %>% mutate(signal = !is.na(`Signal peptide`))
```

Another indicator that a protein is an AMP precursor is the presence of a signal peptide. `r swissprot_precursor %>% filter(signal) %>% nrow` of the `r swissprot_precursor %>% nrow` precursors identified above have well defined signal peptide sequences. The length distribution plot below shows that only precursors longer than about 60 AA are likely to have a signal peptide. Manual inspection of precursors without signal peptides revealed that many are annotated with a pro-peptide indicating that even in this group there is some post-translational processing to produce a mature product. 


```{r, warning=FALSE, message=FALSE}
uniprot_peptides %>% 
  filter(Status=="reviewed") %>% 
  ggplot(aes(x=Length)) + 
  geom_density(aes(colour=signal)) + 
  facet_wrap(~maturity, scales = "free", ncol = 1) + xlim(0,500) +
  labs(x = "Sequence length", y = "Density", colour = "Signal\npeptide")

ggsave("figures/swissprot_signal.png", height = 6, width = 7)
```


SwissProt also includes a small number of larger proteins (>500 AA) that are listed under the keyword Antimicrobial but are very different from classical AMPs. These include some large viral proteins (e.g. [EXLYS_BPDPK](https://www.uniprot.org/uniprot/Q8SCY1)) which show evidence of antibacterial activity but their mode of action and vast difference in size make them outliers from the point of view of building a machine learning model.

### Other peptide datasets

**AmPEP Training Data**

The AmPEP AMP predictor provides its training data available directly for download. The length distribution of sequences in this database is interesting. It shows that sequences classified as AMPs form a clear peak corresponding to mature peptides whereas non-AMP (background) sequences are clearly larger and more likely to represent full length proteins.

```{r}
ampep_data <- read_faa("raw_data/amp_databases/trian_ne_set9894_for_ampep_sever.fasta") %>% add_column(class="non-AMP") %>% 
  rbind(read_faa("raw_data/amp_databases/trian_po_set3298_for_ampep_sever.fasta") %>% add_column(class="AMP")) %>% 
  mutate(length = nchar(seq_aa)) %>% add_column(database="AmPEP")
```

**AMP Scanner v2 Data**

AMP Scanner data used for training, testing and evaluation are available directly for download from [https://www.dveltri.com/ascan/v2/about.html](https://www.dveltri.com/ascan/v2/about.html). In contrast with the AmPEP data the lengths of positive and negative datasets are much more closely matched for AMPScanv2. This reflects specific steps taken by the authors of AMP scanner v2 (detailed in [Veltri et al. 2018](https://doi.org/10.1093/bioinformatics/bty179])) to extract random sub-sequences from non-AMPs that have a matching length distribution to the AMPs themselves. 

```{r}
ampscan_data <- read_faa("raw_data/amp_databases/AMP_Scan2_OrigPaper_Dataset/AMP.tr.fa") %>% add_column(class="AMP") %>% 
  rbind(read_faa("raw_data/amp_databases/AMP_Scan2_OrigPaper_Dataset/DECOY.tr.fa") %>% add_column(class="non-AMP")) %>% 
  mutate(length = nchar(seq_aa)) %>% add_column(database="ampscanner v2")
```


**Xiao et al. Benchmark data**

The benchmark data provided by Xiao et al. 2013 has been used in several studies to provide a somewhat independent estimate of prediction accuracy. Although this data is restricted to sequences less than 100 AA it otherwise resembles the AmPEP data in overall length distribution. This helps explain the extraordinary accuracy of AmPEP when tested with this benchmark. A more worrying issue is that this benchmark, and the AmPEP training data, have length distributions which suggest that positive cases are mature peptides while negative cases are full length proteins. A predictor optimised to perform well on such data will therefore be effective at distinguishing mature peptides from precursor proteins but perhaps not so effective at distinguishing between AMP and non-AMP mature peptides (arguably a more important and interesting task). We would therefore recommend that future work use a negative dataset that is as close as possible to a set of non-AMP mature peptides. Since this is difficult to obtain, the negative data should at least have a similar length distribution to the positive dataset (i.e. reflecting other types of mature peptides).

```{r}

xbench_data <- read_faa("raw_data/benchmarking/datasets/iamp2l/iamp2l_bench.fasta") %>% 
  mutate(class = ifelse(grepl(seq_name,pattern = "^AP"), "AMP", "non-AMP")) %>% 
  mutate(length = nchar(seq_aa)) %>% add_column(database="Xiao et al 2013 benchmark")

ggplot(rbind(ampep_data,ampscan_data,xbench_data),aes(x=length)) + 
  geom_density(aes(color=class)) +
  theme_pubclean() +
  theme(legend.title = element_blank()) + 
  facet_wrap(~database, ncol = 1) +
  labs(x= "Sequence length", y = "Density")
ggsave("figures/amp_predictors_lengths.png", height = 4, width = 8)

```


## Database used for the `ampir` default (precursor) model

Since our goal with `ampir` is to obtain the maximum possible utility for genome-wide scans, we sought to build a positive AMP dataset consisting entirely of precursor proteins. In typical genome-scanning operations this is the only information available.  

To achieve this we used the following criteria to design our database:

Firstly we start with the UniProt database and only include proteins if they are present in either the reviewed or unreviewed elements of this database. Although this removes a small number of proteins from custom AMP databases it allows us to make use of extensive metadata included for all proteins in UniProt. The following filters are then applied:

1. Exclude all mature peptides
2. Exclude unreviewed proteins unless they also appear in APD, DRAMP or dbAMP
3. Remove proteins with lengths < 50 AA since these might be mature peptides included in APD, DRAMP or dbAMP
4. Remove very large proteins (>500 AA) since these are likely to have very different physicochemical properties and are not amenable to prediction by this method
5. Remove identical sequences
6. Remove sequences with nonstandard amino acids

```{r}
uniprot_overlap <- uniprot_peptides %>% 
  mutate(in_apd = (Entry %in% apd$uniprot) | (Sequence %in% apd$seq)) %>% 
  mutate(in_dramp = (Entry %in% dramp$uniprot) | (Sequence %in% dramp$seq)) %>% 
  mutate(in_dbAMP = (Entry %in% dbamp$uniprot) | (Sequence %in% dbamp$seq)) %>% 
  mutate(in_amp_database = in_apd | in_dramp | in_dbAMP)

ampir_db <- uniprot_overlap %>% 
  filter(!is_mature_peptide | is.na(is_mature_peptide)) %>%   # Filtering criterion 1
  filter(Status == "reviewed" | in_amp_database) %>%          # Filtering criterion 2
  filter(Length < 500) %>% filter(Length > 50)  %>%           # Filtering criteria 3 & 4
  distinct(Sequence, .keep_all = TRUE)   %>%                  # Filtering criterion 5
  filter(grepl(Sequence,pattern='^[ARNDCEQGHILKMFPSTWYV]+$')) # Filtering criterion 6
```

This leaves an initial database with `r ampir_db %>% nrow` entries, of which `r ampir_db %>% filter(Status=="unreviewed") %>% nrow` are unreviewed.  


```{r, eval=FALSE}
# Main DB
df_to_faa(ampir_db %>% select(Entry, Sequence) %>% as.data.frame(), "raw_data/amp_databases/ampir_positive.fasta")
```

```bash
cd-hit -i raw_data/amp_databases/ampir_positive.fasta -o raw_data/amp_databases/ampir_positive90.fasta -c 0.90 -g 1
```

```{r}
ampir_90 <- read_faa("raw_data/amp_databases/ampir_positive90.fasta")

ampir_db_summary <- ampir_db %>% mutate(in_90 = (Entry %in% ampir_90$seq_name))
```

As a final step we write the database to a FASTA file and then use `cd-hit` to cluster sequences to 90% identity, keeping only a single representative sequence for each cluster. This reduces the database size (to `r ampir_90 %>% nrow` sequences) but roughly maintains the same length distribution.


Certain organisms are particularly well annotated for AMPs. We find that our final database contains a large number of *Arabidopsis*, human, mouse, chicken and rat sequences. 

```{r}
ampir_db_summary %>% group_by(Organism) %>% 
  summarise(nentries = n(), n90_entries = sum(in_90)) %>% arrange(desc(n90_entries)) %>% head() %>% knitr::kable()
```

### Database files

- Full database (prior to `cd-hit` clustering) along with SwissProt metadata is available at `raw_data/amp_databases/ampir_db.tsv`
- A FASTA formatted file with 90% clustered sequences `raw_data/amp_databases/ampir_positive90.fasta`


```{r, eval=FALSE}
write_tsv(ampir_db,"raw_data/amp_databases/ampir_db.tsv")
```

## `ampir` mature-peptide model

Another approach to AMP prediction is to focus entirely on mature peptides as these are the most likely to have shared/convergent physicochemical properties since they are the active molecules. 

For this approach we build a database as follows:

1. Include all AMPs from the APD, DRAMP and dbAMP databases with lengths >10 AA and < 60 AA
2. Include mature peptides from SwissProt (also with length >10 AA and <60 AA)
3. Remove sequences that are identical or that contain non-standard amino acids

```{r}
ampir_mature <- combined %>% 
  left_join(uniprot_peptides,by = c("uniprot" = "Entry")) %>% 
  filter(length < 60) %>% filter(length > 10) %>% 
  filter(database!="SwissProt" | is_mature_peptide) %>% 
  distinct(seq, .keep_all = TRUE)   %>%                  
  filter(grepl(seq,pattern='^[ARNDCEQGHILKMFPSTWYV]+$')) 
```

The resulting database has a high proportion of peptides

```{r, eval=FALSE}
# Main DB
df_to_faa(ampir_mature %>% select(ID, seq) %>% as.data.frame(), "raw_data/amp_databases/ampir_mature_positive.fasta")
```

Finally cluster these sequences to 90% identity with `cd-hit`

```bash
cd-hit -i raw_data/amp_databases/ampir_mature_positive.fasta -o raw_data/amp_databases/ampir_mature_positive90.fasta -c 0.90 -g 1
```

### Database files

- Full database (prior to `cd-hit` clustering) along with SwissProt metadata is available at `raw_data/amp_databases/ampir_mature_positive.fasta`
- A FASTA formatted file with 90% clustered sequences `raw_data/amp_databases/ampir_mature_positive90.fasta`
